### ISSUE TITLE
Cyclic depencency error: Equations with implicit_dims + SubDomainSet

### ISSUE THREAD

**Issue by rhodrin** [opesci/devito#827]
Take for example the following code:

```
import numpy as np
from math import floor
from devito import (Grid, Function, TimeFunction, Eq, solve, Operator, SubDomainSet,
                    Dimension)

n_domains = 10

class Inner(SubDomainSet):
    name = 'inner'

bounds_xm = np.zeros((n_domains,), dtype=np.int32)
bounds_xM = np.zeros((n_domains,), dtype=np.int32)
bounds_ym = np.zeros((n_domains,), dtype=np.int32)
bounds_yM = np.zeros((n_domains,), dtype=np.int32)

for j in range(0, n_domains):
    bounds_xm[j] = j
    bounds_xM[j] = n_domains-1-j
    bounds_ym[j] = floor(j/2)
    bounds_yM[j] = floor(j/2)

bounds = (bounds_xm, bounds_xM, bounds_ym, bounds_yM)

inner_sd = Inner(N=n_domains, bounds=bounds)

grid = Grid(extent=(10, 10), shape=(10, 10), subdomains=(inner_sd, ))

f = TimeFunction(name='f', grid=grid, dtype=np.int32)
f.data[:] = 0

v = Function(name='v', dimensions=(inner_sd.implicit_dimension, ), shape=(n_domains, ))

for j in range(1, n_domains):
    v.data[j] = v.data[j-1]+1.0
    
idims = (grid.stepping_dim, inner_sd.implicit_dimension, 
         inner_sd.dimensions[0], inner_sd.dimensions[1])

stencil = Eq(f.forward, v*solve(Eq(f.dt, 1), f.forward),
            subdomain=grid.subdomains['inner'], implicit_dims=idims)

op = Operator(stencil, dle='noop')
op(time_m=0, time_M=1, dt=1)
```

In the above we are trying to set `v` to a different value in each subdomain of the `SubDomainSet`. This 'sometimes' works (as expected) but other times throws and error of the form:
```
ValueError: A cyclic dependency exists amongst OrderedDict([(n, {n, time}), (time, {n})])
```

Note that this was previously untested functionality and therefore there's no reason it should have worked 'out of the box', but the above issue needs to be fixed.

**Comment by FabioLuporini** [comment]
I wonder whether this is still an issue

**Comment by rhodrin** [comment]
I checked fairly recently and it still was. Will check again when I get back.

**Comment by rhodrin** [comment]
This is still an issue by the way. I'm going to start looking into it now.

**Comment by rhodrin** [comment]
Ok, this was occurring due to the `implicit_dimensions` sometimes being passed incorrectly to the implicit equations which was in turn occurring due to the 'implicit dimension' being included twice and sometimes (due to the nature of `_.free_symbols`) in the incorrect order. PR #965 fixes this.

**Comment by FabioLuporini** [comment]
fixed by #965 .
